### PR TITLE
Simplify implementation.

### PR DESCRIPTION
**Pull request by patrickt**
This code, like so much TH, is sort of a rat’s nest, but I took some refactoring passes through it and simplified some pattern matching and context handling. Pinging @robrix to see if he has any thoughts on it.

### CODE REVIEW

**Comment by kutyel** [comment]
Hi Patrick! Have you considered [`th-abstraction`](https://hackage.haskell.org/package/th-abstraction)? 😄

File: src/Control/Effect/TH.hs [hunk 1]
```diff
@@ -63,32 +64,47 @@ data PerDecl = PerDecl
...
   fmap join . for names $ \ctorName -> do
-    let downcase = \case
-          x : xs -> toLower x : xs
-          [] -> []
-        functionName = TH.mkName . downcase . TH.nameBase $ ctorName
-    let decl = PerDecl {..}
+    let downcase =
+          mkName . \case
+            x : xs -> toLower x : xs
+            [] -> []
```
Comment (kutyel) [inline src/Control/Effect/TH.hs]: `LambdaCase` sure it is nice, but to me it reads a bit strange here 🤔 
```suggestion
    let downcase (x : xs) = mkName (toLower x : xs)
        downcase [] = mkName []
```

**Comment by patrickt** [comment]
@kutyel I just learned about `th-abstraction` — it looks pretty good! I ran into problems trying to use third-party libraries from TH splices using `ghcide`, but that might not happen with `th-abstraction`, so I think I might take a run at it in a separate PR.

**Comment by kutyel** [comment]
Glad to help! 😇

**Pull request merged**